model/src/test/kotlin/ScannerOptionsTest.kt

```diff
@@ -0,0 +1,405 @@
+        putStringOption(key = "some-option", value = "someValue")
+        putThresholdOption(key = "max-count", value = 75.0)
+        putStringOption(key = "another-option", value = "foo", relevant = false)
+    }
+    val compatibleSubOptions2 = SubOptions.create {
+        putStringOption(key = "some-option", value = "someValue")
+        putThresholdOption(key = "max-count", value = 50.0)
+    }
+    val compatibleOption1 = factory(compatibleSubOptions1)
+    val compatibleOption2 = factory(compatibleSubOptions2)
+    val options1 = ScannerOptions(setOf(compatibleOption1))
+
+    compatibleOption2.isCompatible(options1) shouldBe true
+
+    // check compatibility to a missing option
+    val unmatchedOption = factory(SubOptions(jsonMapper.createObjectNode()))
+    val emptyOptions = ScannerOptions(emptySet())
+    unmatchedOption.isCompatible(emptyOptions, strict = false) shouldBe false
+
+    // check compatibility to different option in strict mode
+    val incompatibleSubOptions1 = SubOptions.create {
+        putStringOption("foo")
+    }
+    val incompatibleSubOptions2 = SubOptions.create {
+        putStringOption("bar")
+    }
+    val incompatibleOption1 = factory(incompatibleSubOptions1)
+    val incompatibleOption2 = factory(incompatibleSubOptions2)
+    val options2 = ScannerOptions(setOf(incompatibleOption1))
+
+    incompatibleOption2.isCompatible(options2) shouldBe false
+
+    // check compatibility to different option in lenient mode
+    incompatibleOption2.isCompatible(options2, strict = false) shouldBe true
+}
+
+/**
+ * Return a [ScannerOptions] instance that includes all supported options.
+ */
+private fun completeOptions(): ScannerOptions {
+    val outputOption = OutputFormatOption(SubOptions.create { putStringOption("xml") })
+    val copyrightOption = CopyrightResultOption(
+        SubOptions.create { putStringOption(key = "consolidate", value = "true") }
+    )
+    val emailOption = EmailResultOption(
+        SubOptions.create { putThresholdOption(100.0) }
+    )
+    val licenseOption = LicenseResultOption(
+        SubOptions.create {
+            putStringOption(key = "consolidate", value = "true")
+            putStringOption(key = "license-text", value = "true")
+            putThresholdOption(key = "license-score", value = 75.0)
+        }
+    )
+    val metadataOption = MetadataResultOption(
+        SubOptions.create { putStringOption(key = "mark-source", value = "true") }
+    )
+    val packageOption = PackageResultOption(
+        SubOptions.create { putStringOption(key = "consolidate", value = "true") }
+    )
+    val urlOption = UrlResultOption(
+        SubOptions.create { putThresholdOption(key = "max-url", value = 64.0) }
+    )
+
+    return ScannerOptions(
+        setOf(
+            outputOption,
+            copyrightOption,
+            emailOption,
+            licenseOption,
+            metadataOption,
+            packageOption,
+            urlOption
+        )
+    )
```

Comment:
You partly go towards a DSL with the `SubOptions.create` function, did you consider adding more DSL functions so that you could write this as:
```
scannerOptions {
  outputFormatOption {
    stringOption("xml")
  }

  copyrightResultOption {
    stringOption(key = "consolidate", value = "true")
  }

  emailOption {
    thresholdOption(100.0)
  }

  licenseOption {
    stringOption(key = "consolidate", value = "true")
    stringOption(key = "license-text", value = "true")
    thresholdOption(key = "license-score", value = 75.0)
  }

  ...
}
```

Reply:
I was actually thinking about adding more DSL-like features, but then I am not sure whether this is worth the effort. The main consumers of this DSL are concrete scanner implementations, and their number is limited. But I agree, this is looking very nice.